### PR TITLE
draft of design doc process

### PR DESCRIPTION
**Pull request by elindsey**
This is a draft of a lightweight design/review process. Some aspects were cribbed from Java's JEP process and Cockroach's RFCs. It includes two prior examples in the same spirit (though not using the exact same template).

Roughly, these are my principles that went into the draft:
1. As little process as necessary. More structure can be added later if it's useful, but best to arrive there gradually and grudgingly.
2. Deep ownership by the author
3. Explicit communication, especially around things we are choosing _not_ to do

It's opinionated, and I'd like to hear everyone else's opinions/input as well so we can land on something useful. Comment away! 🙂

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5988)

### CODE REVIEW

File: doc/developer/design/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+3. If there are multiple alternative implementations and no clear best option
+4. If it's going to involve multiple people coordinating changes
+
+Many smaller changes do still benefit from a quick design doc to clarify
+thinking.
+
+## How should you make a design document?
+
+1. Copy the template to a new date-prefixed file in `doc/developer/design` and fill it in.
+2. Submit a pull request - this makes it easy for others to add writtne comments.
```
Comment (quodlibetor) [inline doc/developer/design/README.md:50]: typo: `writtne`

File: doc/developer/design/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+3. Schedule a meeting, typically in the 30-60min range. Provide time to read
```
Comment (quodlibetor) [inline doc/developer/design/README.md:51]: I would be tempted to say that if you think it needs to be more than 60 minutes there should be separate meetings about clear parts of it.

**Review by quodlibetor — APPROVED** [review]
Looks great!

**Review by cirego — APPROVED** [review]
This doc is clear and to the point. I like it!

File: doc/developer/design/00000000_template.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+# Non-Goals
+
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+
+# Description
+
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
```
[the remark this answers is not in the document]
Comment (elindsey) [reply doc/developer/design/00000000_template.md:17]: I usually expect this to be in summary/goals - those sections should have some notion of customer impact to answer the "why are we even doing this" part. I'll update the template with some mention of that, referencing draft changelogs/press releases as potential ways of framing it

File: doc/developer/design/README.md (hunk 1 above)
Comment (elindsey) [reply doc/developer/design/README.md:50]: Thanks!

File: doc/developer/design/README.md (hunk 2 above)
Comment (elindsey) [reply doc/developer/design/README.md:51]: Good point - at that point I think it's also an indication that the single design doc is trying to do too much, and you're better off writing a more focused doc for a subcomponent. I'll add mention of both to this section

**Review by benesch — APPROVED** [review]
This seems great to me!

File: doc/developer/design/00000000_template.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Summary
```
[the remark this answers is not in the document]
Comment (elindsey) [reply doc/developer/design/00000000_template.md:1]: Good point! https://github.com/MaterializeInc/materialize/pull/6065

**Pull request merged**